inc/saiprefixcompression.h

```diff
@@ -0,0 +1,305 @@
+ *
+ *    Microsoft would like to thank the following companies for their review and
+ *    assistance with these files: Intel Corporation, Mellanox Technologies Ltd,
+ *    Dell Products, L.P., Facebook, Inc., Marvell International Ltd.
+ *
+ * @file    saiprefixcompression.h
+ *
+ * @brief   This module defines SAI prefix compression interface
+ */
+
+#if !defined (__SAIPREFIXCOMPRESSION_H_)
+#define __SAIPREFIXCOMPRESSION_H_
+
+#include <saitypes.h>
+
+/**
+ * @defgroup SAIPREFIXCOMPRESSION SAI - Prefix Compression API definitions
+ *
+ * @{
+ */
+
+/**
+ * @brief Attribute Id for SAI prefix compression object
+ */
+typedef enum _sai_prefix_compression_table_attr_t
+{
+    /**
+     * @brief Start of attributes
+     */
+    SAI_PREFIX_COMPRESSION_TABLE_ATTR_START,
+
+    /**
+     * @brief Label attribute used to unique identify Table.
+     *
+     * @type char
+     * @flags CREATE_AND_SET
+     * @default ""
+     */
+    SAI_PREFIX_COMPRESSION_TABLE_ATTR_LABEL = SAI_PREFIX_COMPRESSION_TABLE_ATTR_START,
+
+    /**
+     * @brief Prefix Compression table default entry BINCODE.
+     * BINCODE applied to any address that does not match a prefix in the table
+     *
+     * @type sai_uint32_t
+     * @flags MANDATORY_ON_CREATE | CREATE_AND_SET
+     */
+    SAI_PREFIX_COMPRESSION_TABLE_ATTR_DEFAULT_ENTRY_BINCODE,
+
+    /**
+     * @brief End of attributes
+     */
+    SAI_PREFIX_COMPRESSION_TABLE_ATTR_END,
+
+    /** Custom range base value */
+    SAI_PREFIX_COMPRESSION_TABLE_ATTR_CUSTOM_RANGE_START = 0x10000000,
+
+    /** End of custom range base */
+    SAI_PREFIX_COMPRESSION_TABLE_ATTR_CUSTOM_RANGE_END
+
+} sai_prefix_compression_table_attr_t;
+
+/**
+ * @brief Attribute Id for SAI prefix compression object
+ */
+typedef enum _sai_prefix_compression_entry_attr_t
+{
+    /**
+     * @brief Start of attributes
+     */
+    SAI_PREFIX_COMPRESSION_ENTRY_ATTR_START,
+
+    /**
+     * @brief Prefix Compression entry BINCODE
+     *
+     * @type sai_uint32_t
+     * @flags MANDATORY_ON_CREATE | CREATE_AND_SET
+     */
+    SAI_PREFIX_COMPRESSION_ENTRY_ATTR_BINCODE = SAI_PREFIX_COMPRESSION_ENTRY_ATTR_START,
+
+    /**
+     * @brief End of attributes
+     */
+    SAI_PREFIX_COMPRESSION_ENTRY_ATTR_END,
+
+    /** Custom range base value */
+    SAI_PREFIX_COMPRESSION_ENTRY_ATTR_CUSTOM_RANGE_START = 0x10000000,
+
+    /** End of custom range base */
+    SAI_PREFIX_COMPRESSION_ENTRY_ATTR_CUSTOM_RANGE_END
+
+} sai_prefix_compression_entry_attr_t;
+
+/**
+ * @brief Prefix Compression entry
+ */
+typedef struct _sai_prefix_compression_entry_t
+{
+    /**
+     * @brief Switch ID
+     *
+     * @objects SAI_OBJECT_TYPE_SWITCH
+     */
+    sai_object_id_t switch_id;
+
+    /**
+     * @brief Prefix Compression Table ID
+     *
+     * @objects SAI_OBJECT_TYPE_PREFIX_COMPRESSION_TABLE
+     */
+    sai_object_id_t prefix_table_id;
+
+    /**
+     * @brief IP Prefix Destination
+     */
+    sai_ip_prefix_t prefix;
```

Comment:
VRF is not being considered here. Is this intentional ?

Reply:
Yes. Prefix Compression is a table of longest prefix match. The entries in the table should indicate the long prefix which are mapped to a unique value. This should not be confused with sairoute.cpp or such features. This has to standalone outside the forwarding table in order not to affect the forwarding path.

Reply:
@helloanandhi, do you still want to discuss this after the discussion we had about this feature on 8/8/2024 meeting?

Reply:
@nadershinouda,  these bincodes are applied only to routed packets or also to layer 2 bridged packets ?

Reply:
@helloanandhi This is only dealing with L3 routed packets. The table entries are prefix/mask of addresses that get mapped to a meta (bincode) value. The idea is to add longest prefix addresses as entries to this table, that get mapped to this meta value. Packets would need to have either IPV4 address or IPV6 address for ACL to be able to match on the meta value

If this is not clear I don't mind adding it to the list of things that we need to discuss.

Reply:
@nadershinouda , It is clear. Thanks for the response.

Reply:
@helloanandhi can I resolve this conversation?

Reply:
I think the question is not whether the packet is IP or non-IP, rather more on the forwarding operation. Let us say, the switch receives an IP packet with DIP as 10.1.1.3 on a VLAN. The DMAC of the packet does not match the router's MAC and the packet will get bridged. 
Now a prefix-compression entry for 10.1.1.x to assign meta-data X should still apply for this packet and meta-data X should be associated with the bridged packet. Is this true?

Reply:
the packet would be dropped before it reaches your ACL, in that case the packet will not be affected. If the packet reaches the ACL for any reason, prefix compression only applies to IPV4/IPV6 rules, so it would not apply to this packet. 

I can request for time next week for us to discuss this if needed. Might be easier over a meeting than on this review

Reply:
Hi Nader,

Can you please explicitly clarify whether the below statements are true. 

"Prefix compression applies to a packet with IP header irrespective of whether the packet undergoes L2 forwarding or L3 forwarding. For e.g., a packet with IP header that undergoes bridging operation, will still get the metadata assigned by the prefix compression table".